### ISSUE TITLE
ArrayIndexOutOfBoundsException in ReadGrammar.java:210

### ISSUE THREAD

**Issue by mhuesch** [GaloisInc/ppaml-cp5#19]
```
peval@ppamlsys:/tmp/peval.kDx6dC.CRASH/evaluator$ ./eval.sh ../result/ ../ground_truth/ ../output/   
Symbol sets from pcfgla.grammar and pcfgla.states match.
Reading grammar ...
Calling with { -gr => /tmp/peval.kDx6dC.CRASH/result/pcfgla.grammar -lex => /tmp/peval.kDx6dC.CRASH/result/pcfgla.lexicon -st => /tmp/peval.kDx6dC.CRASH/result/pcfgla.states.fixed -out => pcfgla.gr }
Read 95 states.
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: 3
        at edu.berkeley.nlp.PCFGLA.ReadGrammar.readGrammar(ReadGrammar.java:210)
        at edu.berkeley.nlp.PCFGLA.ReadGrammar.main(ReadGrammar.java:360)
reading grammar exited with return code 1
```
I see the above message when trying to evaluate these results:
[result.zip](https://github.com/GaloisInc/ppaml-cp5/files/63342/result.zip)

Hehe, do you know what might cause this index out of bounds exception? It is not obvious to me from looking at the source. Thanks!

**Comment by hhexiy** [comment]
Hi Michael,

It seems the grammar file has state annotation starting from 1; however,
the program expects 0-based indexing for state.  The output_format.txt says
it “must be numbered from 0 up to N
(exclusive) consecutively, where N is the total number of states of the
original tag.”

For example, the number of states of the non-terminal @QP^g is 3 in the
.state file, thus valid annotation of this non-terminal should be @QP^g_0,
@QP^g_1, @QP^g_2, but in the .grammar file there is @QP^g_3.

He

**Comment by mhuesch** [comment]
Ah great, nice find! I will contact the team and have them check on a one-indexing vs zero-indexing bug.

Thanks!

**Comment by mhuesch** [comment]
The team has made the change, and I am now seeing a new out of bounds error:
```
peval@ppamlsys:/tmp/peval.glBnVU.CRASH/evaluator$ ./eval.sh ../result/ ../ground_truth/ ../output/
Symbol sets from pcfgla.grammar and pcfgla.states match.
Reading grammar ...
Calling with { -gr => /tmp/peval.glBnVU.CRASH/result/pcfgla.grammar -lex => /tmp/peval.glBnVU.CRASH/result/pcfgla.lexicon -st => /tmp/peval.glBnVU.CRASH/result/pcfgla.states.fixed -out => pcfgla.gr }
Read 76 states.
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: 1
        at edu.berkeley.nlp.PCFGLA.ReadGrammar$MyBinaryRule.createBinaryRule(ReadGrammar.java:82)
        at edu.berkeley.nlp.PCFGLA.ReadGrammar.readGrammar(ReadGrammar.java:205)
        at edu.berkeley.nlp.PCFGLA.ReadGrammar.main(ReadGrammar.java:360)
reading grammar exited with return code 1
```
Files here: 
[result.zip](https://github.com/GaloisInc/ppaml-cp5/files/67356/result.zip). Do you know what's going on here?

It seems like there are some invariants here that we should be checking before we start running the solutions (to reduce confusion and me having to pester you all the time). I'll make a list of the errors that have come up and maybe add some checking to the validation script.

**Comment by hhexiy** [comment]
Hi Michael,

It doesn't seem they fixed the problem. @QP^g has 3 states, but the grammar
file still has grammars like QP^g_2 -> *@QP^g_3* JJ_1 0.00196078.

It seems like there are some invariants here that we should be checking
> before we start running the solutions (to reduce confusion and me having to
> pester you all the time). I'll make a list of the errors that have come up
> and maybe add some checking to the validation script.
>

Yes, I agree. I can also try to add more informative error message in the
code once you have the error list.

On Fri, Dec 18, 2015 at 10:30 PM, Michael Hueschen <notifications@github.com

**Comment by mhuesch** [comment]
I am seeing that `@QP^g` has 2 states, but appears in the grammar with indices 0,1, and 2. But, still the same idea. I will follow up with the team about this. Thanks!

I will also open another issue about improving validation, and list the commonly violated invariants there.
